lib/package_config.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A package configuration is a way to assign file paths to package URIs,
+/// and vice-versa,
+library package_config.package_config;
+
+import "dart:io" show File, Directory;
+import "src/discovery.dart" as discover;
+import "src/package_config.dart";
+import "src/package_config_json.dart";
+
+export "src/package_config.dart" show PackageConfig, Package;
+export "src/discovery_analysis.dart" show PackageContext;
+
+/// Reads a specific package configuration file.
+///
+/// The file must exist and be readable.
+/// It must be either a valid `package_config.json` file
+/// or a valid `.packages` file.
+/// It is considered a `package_config.json` file if its first character
+/// is a `{`.
+///
+/// If the file is a `.packages` file, first checks if there is a
+/// `.dart_tool/package_config.json` file next to the original file,
+/// and if so, loads that instead.
+///
+/// If [extraData] is provided, and the loaded file is a `package_config.json`
+/// file, then any unknown JSON object entries in the main JSON object of the
+/// file are added to [extraData] with the same name.
+PackageConfig loadPackageConfig(File file,
+        [Map<String, dynamic> /*?*/ extraData]) =>
```

Comment:
It's reasonable. I tried to keep the extraData as separate from the actual data, because it's not *part of* that data, it's just stored in the same file alongside it.
It'll be a better API if there is an optional `extraData` on both `PackageConfig` and `Package`, and you can still pass a boolean to control whether it's parsed or not.

Reply:
Another reason is that the resulting map is mutable, but the created class is immutable. I wanted to either make the map immutable, which is costly, or keep it outside the immutable object.
I'll go for the former instead.